### ISSUE TITLE
On Build, program runs

### ISSUE THREAD

**Issue by KarlSevenSeven** [WaynePhillipsEA/twinbasic#873]
**Describe the bug**
The code runs on Build.
Evidence is the MsgBox that pops up.

**To Reproduce**
1. Paste this:
```
Attribute VB_Name = "Play"
Option Explicit

Public Declare Function PlaySound Lib "winmm.dll" Alias "PlaySoundA" _
    (ByVal lpszName As String, _
     ByVal hModule As Long, _
     ByVal dwFlags As Long) As Long

Private Const SND_APPLICATION As Long = &H80
Private Const SND_ALIAS As Long = &H10000
Private Const SND_ALIAS_ID As Long = &H110000
Private Const SND_ASYNC As Long = &H1
Private Const SND_FILENAME As Long = &H20000
Private Const SND_LOOP As Long = &H8
Private Const SND_MEMORY As Long = &H4
Private Const SND_NODEFAULT As Long = &H2
Private Const SND_NOSTOP As Long = &H10
Private Const SND_NOWAIT As Long = &H2000
Private Const SND_PURGE As Long = &H40
Private Const SND_RESOURCE As Long = &H40004
Private Const SND_SYNC As Long = &H0

Sub Main()

Dim File As String
Dim k As String

'k = Command$
'File = k

File = "C:\Privat\Sound\45.WAV"

MsgBox File

WAVPlay File

End Sub


Sub WAVPlay(ByRef File As String)
        
PlaySound File, 0, SND_FILENAME

End Sub
```

3. Click on Build.

**Expected behavior**
Making the EXE but not run the code.

**Desktop (please complete the following information):**
 - OS: Windows 10 GERMAN
 - twinBASIC beta 47

**Comment by WaynePhillipsEA** [comment]
There is a `Run After Build` project option to turn this feature off.

**Comment by KarlSevenSeven** [comment]
Aha ;-)
Then it's not a bug.
I think this should be off by default as it is unexpected.

**Comment by WaynePhillipsEA** [comment]
Somewhere we discussed that it might be better to have two build buttons instead.  One for just building, and one for building & running.   But we need a modified icon for the build&run option which I haven't got around to yet.

**Comment by KarlSevenSeven** [comment]
I don't understand what it is useful for.
We can run the built EXE manually.
I would like an option/button/menu/whatever "Open Build folder" better.
Perhaps in the menu between Build and Clean.
Just my 0.02$

**Comment by WaynePhillipsEA** [comment]
> I don't understand what it is useful for.
> We can run the built EXE manually.

It is useful for saving all those unnecessary extra clicks!   I for one, use it a lot.

**Comment by KarlSevenSeven** [comment]
Then I'm with you!

Having an Open Build folder option would avoid extra clicks as well.

**Comment by WaynePhillipsEA** [comment]
Yes, I think an `Open Build Folder` command of some sort is a good idea.

**Comment by EduardoVB** [comment]
> It is useful for saving all those unnecessary extra clicks! I for one, use it a lot.

For me 1/100 times I need to run the program after build.

While we are debugging maybe if we are trying to debug something that we think it might not behave the same as in the IDE.
In most of case VBx developers trust that the exe will do the same as the program does in the IDE, so why the need to run it after build?

**Comment by WaynePhillipsEA** [comment]
> > It is useful for saving all those unnecessary extra clicks! I for one, use it a lot.
> 
> For me 1/100 times I need to run the program after build.
> 
> While we are debugging maybe if we are trying to debug something that we think it might not behave the same as in the IDE. In most of case VBx developers trust that the exe will do the same as the program does in the IDE, so why the need to run it after build?

Well, it could have something to do with twinBASIC being very much a BETA compiler, that you absolutely **should not** trust the compiled EXE does match the IDE/debugger version 😁

**Comment by KarlSevenSeven** [comment]
> that you absolutely should not trust the compiled EXE
😆

Let us select for options after build.

- Do nothing
- Run (+ parameters)
- Run something to configure, eg zip
- Open folder

**Comment by WaynePhillipsEA** [comment]
@KarlSevenSeven that sounds like a good idea.

**Comment by WaynePhillipsEA** [comment]
Perhaps we could make it even more flexible than that, and allow you to launch a procedure (in the IDE) after a successful build.  So you can then write a build script that does whatever else with the built executable (sign it, zip it, launch InnoSetup etc).

**Comment by bclothier** [comment]
Just to note, though, such native build script implies a good robust API for piping external commands so that it's easy to construct a script that uses external programs for the build and get good error reporting, hopefully without lot of boilerplate code; somewhere between the `ShellAndWait` and `Wscript.Exec`.

**Comment by WaynePhillipsEA** [comment]
@bclothier adding an option to execute a procedure after a successful build will take only a few minutes, but of course enhancing `Shell` to allow for streaming of the outputs will take more work.   So the question becomes, shall we add the new setting first, or wait for the enhanced `Shell`?

I believe the Run-After-Build procedure execution could be very handy (in multiple different scenarios), and so my thought is to implement that now, then we can look at the `Shell` enhancements separately.

We could probably do away with the setting and instead use an attribute instead.
```
@RunAfterBuild
Private Sub MyBuildScript()
   ' ...
End Sub
```

**Comment by bclothier** [comment]
To be clear, I consider the enhanced shell a separate request and not one that requires waiting. I only wanted to comment because that will be inevitably asked for. 

RE: attribute; I don't know. I usually consider build script a separate compilation unit and it should be able to be compiled independently of the program being built, and wouldn't become a part of the shipped output normally. Sure, `@RunAfterBuild` could imply that it won't get compiled into the output but being in the same compilation unit doesn't feel right to me.

**Comment by WaynePhillipsEA** [comment]
Sometimes it seems that I am the one from another solar system 😁

Honestly, I'm struggling to see where the complexity arises.   If you want to run a batch file after each build, you might do this:

```vb
@RunAfterBuild
Private Sub MyBuildScript()
   Shell "path_to_script.bat"
End Sub
```

If you want to launch the generated EXE after each build for testing, then you do this:

```vb
@RunAfterBuild
Private Sub MyBuildScript()
   Shell App.BuildPath   ' not a real property yet of course...
End Sub
```

If you want to do something more complicated, then fine:

```vb
@RunAfterBuild
Private Sub MyBuildScript()
   Shell "sign_my_app " & App.BuildPath
   TwinBasicZevenZip.CreateArchive App.BuildPath, App.BuildPath & ".zip"
   MyFtpClient.UploadFile App.BuildPath & ".zip"
End Sub
```

The choice is yours.  It's flexible.  **_It's optional._**  It's unobtrusive (doesn't affect the generated EXE).  It's integrated.  Seems like a no-brainer to me.

Sure, one-day when we support multiple-project loading in the IDE, it would definitely make sense to segregate it and have any build events code in a separate project.  But right now, just being able to trigger a procedure after a successful build seems very useful.

**Comment by bclothier** [comment]
I think it's because I can't think of another ecosystem where the build subsystem is also integrated into the project being built. The build always has been a separate tool and usually follows the compiler, which might be why I stumbled on the idea a bit. 

I do have to admit that I can't argue against the simplicity, and the scenarios where the build script must be a separate thing isn't a thing right now so we can always expand the system when we support multiple projects or need to share the build steps for common build tasks. 

I think that because it's easy to set up, it can't hurt to experiment and see how it goes. If it doesn't work out, we can go back to the drawing board.

**Comment by EduardoVB** [comment]
Self building technology Ⓒ 🤣

**Comment by wqweto** [comment]
> I think it's because I can't think of another ecosystem where the build subsystem is also integrated into the project being built. 

Most modern languages do this. For instance in Zig the project file (i.e. the vbp) is actually an executable zig script which can do whatever pre and post-build tasks are needed (like codesigning etc).

**Comment by WaynePhillipsEA** [comment]
In BETA 49 the project setting of `Run After Build` has now been removed.  As discussed, we've now got a much more flexible `[ RunAfterBuild ]` attribute that can be applied to a procedure in a standard module, like so:

```vb
[ RunAfterBuild ]
Private Sub BuildScript()
    Shell App.LastBuildPath
End Sub
```

`App.LastBuildPath` is also new, and populated immediately after the build completes (and before BuildScript runs).  The new auto-increment version feature (#854) only kicks in ***after*** your build script completes (so that the App object version numbers are valid according to the just built executable).

If you just want to open the build output folder in Windows Explorer, you can use something like this:

```vb
[ RunAfterBuild ]
Private Sub BuildScript()
    Dim buildFolder As String = Left$(App.LastBuildPath, InStrRev(App.LastBuildPath, "\"))
    Shell "explorer """ & buildFolder & """", vbNormalFocus
End Sub
```

**Comment by WaynePhillipsEA** [comment]
I'm going to close this issue as I believe the issue has been resolved by the new feature.  Problems involving the new `[ RunAfterBuild ]` attribute should be opened as separate issues.  Thanks!

**Comment by WaynePhillipsEA** [comment]
Yes, the compiler allows only one per project.

**Comment by wqweto** [comment]
Btw, `RunBeforeBuild` would allow devising code generators for custom DSLs (e.g. some form of compiling grammars etc.)

**Comment by bclothier** [comment]
I think for code generators, the event need to be earlier so that it's usable during the development. `RunBuildBefore` means we are using generated code form the last build, but not necessarily the one when we edit the template. That can lead to a subpar coding experience. Something like `BuildAfterEdit` (only for this module, obviously) so that the generated code is immediately updated & available in the rest of the project.

**Comment by WaynePhillipsEA** [comment]
I'm not necessarily against something like a `RunAfterEdit` event, I would primarily be concerned about the performance impact.  As such, I'd want to see some example use cases so that we can think of ways to minimize that performance impact.

**Comment by bclothier** [comment]
Obviously, it can't be after every keystroke. 😄 

Maybe after saving the changes to the marked module is a better option, and only if it compiles fully (no sense in running if it can't compile). Hopefully it shouldn't be often, and one way we can help reduce cascading errors is to retain a backup copy of the original generated code for the purpose of references so that we don't actually replace the code  until there is a valid & well-formed generated module to replace and minimizing the errors in the referencing modules due to the changes made.

This should be no different from what we experience currently when we edit a form and rename a control that has several code reference to it, no?

**Comment by EduardoVB** [comment]
Shouldn't these kind of things be in something like an add-in environment?

**Comment by bclothier** [comment]
No, not necessarily. Maybe the code generator itself can be provided as an add-in but the template used for the code generating would typically be a part of the consuming project and in theory should be editable, similar to the forms.

**Comment by EduardoVB** [comment]
Doing something after every source code change seems to me more something like an add-in to handle, than the program itself.

I'm not saying I'm against the idea, but I would like to see a sample use too.

**Comment by EduardoVB** [comment]
Once there is an add-in environment, it could add the possibility to run a Sub from the source project on any add-in event, like @RunOnAddInEvent("EventName") , including @RunOnAddInEvent("AfterBuild")
